lib/wasix/src/runtime/module_cache/types.rs

```diff
@@ -127,50 +128,122 @@ impl CacheError {
     }
 }
 
-/// The XXHash hash of a WebAssembly module.
+/// The hash of a WebAssembly module.
 #[derive(Debug, Copy, Clone, PartialEq, Eq, Hash, PartialOrd, Ord)]
-pub struct ModuleHash([u8; 8]);
+pub enum ModuleHash {
+    XXHash([u8; 8]),
+    Sha256([u8; 32]),
+}
 
 impl ModuleHash {
-    /// Create a new [`ModuleHash`] from the raw XXHash hash.
-    pub fn from_bytes(key: [u8; 8]) -> Self {
-        ModuleHash(key)
+    /// Create a new [`ModuleHash`] from the raw xxhash hash.
+    pub fn xxhash_from_bytes(key: [u8; 8]) -> Self {
+        Self::from_bytes(key)
+    }
+
+    /// Create a new [`ModuleHash`] from the raw sha256 hash.
+    pub fn sha256_from_bytes(key: [u8; 32]) -> Self {
+        Self::from_bytes(key)
+    }
+
+    fn from_bytes<const N: usize>(key: [u8; N]) -> Self {
+        if N == 8 {
+            let key = key.as_slice().try_into().unwrap();
+            ModuleHash::XXHash(key)
+        } else if N == 32 {
+            let key = key.as_slice().try_into().unwrap();
+            ModuleHash::Sha256(key)
+        } else {
+            panic!("Only keys with size 8 or 32 are accepted")
+        }
+    }
```

Comment:
This function is not proper. We send a different type depending on the length, which is highly confusion and can lead to many errors